applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_structural_response_function.cpp

```diff
@@ -0,0 +1,942 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Martin Fusseder, https://github.com/MFusseder 
+//   
+
+
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/openmp_utils.h"
+#include "adjoint_structural_response_function.h"
+
+
+// Application includes
+
+namespace Kratos
+{
+
+    /// Constructor.
+    AdjointStructuralResponseFunction::AdjointStructuralResponseFunction(ModelPart& rModelPart, Parameters& rParameters)
+      : mrModelPart(rModelPart)
+    {
+        KRATOS_TRY;
+
+        mSensitivityModelPartName =
+           rParameters["sensitivity_model_part_name"].GetString();
+
+        Parameters nodal_sensitivity_variables = rParameters["nodal_sensitivity_variables"];
+        mNodalSensitivityVariables.resize(nodal_sensitivity_variables.size());
+        for (unsigned int i = 0; i < nodal_sensitivity_variables.size(); ++i)
+            mNodalSensitivityVariables[i] = nodal_sensitivity_variables[i].GetString();
+
+        Parameters element_sensitivity_variables = rParameters["element_sensitivity_variables"];
+        mElementSensitivityVariables.resize(element_sensitivity_variables.size());
+        for (unsigned int i = 0; i < element_sensitivity_variables.size(); ++i)
+            mElementSensitivityVariables[i] = element_sensitivity_variables[i].GetString();
+
+        Parameters condition_sensitivity_variables = rParameters["condition_sensitivity_variables"];
+        mConditionSensitivityVariables.resize(condition_sensitivity_variables.size());
+        for (unsigned int i = 0; i < condition_sensitivity_variables.size(); ++i)
+            mConditionSensitivityVariables[i] = condition_sensitivity_variables[i].GetString();
+
+        // Set gradient mode
+        const std::string gradient_mode = rParameters["gradient_mode"].GetString();
+
+        // Mode 1: semi-analytic sensitivities
+        if (gradient_mode== "semi_analytic")
+        {
+            mGradientMode = 1;
```

Comment:
I don't have an overall picture of how this works, but it seems strange to me that you specify the gradient_mode in this class. Is it not the element that needs to check this? Can this class be independent of this setting?